AutoCollection/StatsBeat.ts

```diff
@@ -0,0 +1,218 @@
+import Config = require("../Library/Config");
+import Context = require("../Library/Context");
+import { MetricTelemetry } from "../Declarations/Contracts";
+
+const STATS_BEAT_LANGUAGE = "node";
+const STATS_BEAT_IKEY = "";
+
+
+export class StatsBeat {
+
+    // Counters
+    private _lastRequests: { totalRequestCount: number; time: number; };
+    private _totalRequestCount: number = 0;
+    private _totalSuccesfulRequestCount: number = 0;
+    private _totalFailedRequestCount: number = 0;
+    private _retryCount: number = 0;
+    private _exceptionCount: number = 0;
+    private _throttleCount: number = 0;
+    private _intervalRequestExecutionTime: number = 0;
+    private _lastIntervalRequestExecutionTime: number = 0;
+
+    private _collectionInterval: number = 900000; // 15 minutes
+    private _client: TelemetryClient;
+    private _handle: NodeJS.Timer | null;
+    private _isEnabled: boolean;
+    private _isInitialized: boolean;
+    private _config: Config;
+
+    // Custom dimensions
+    private _resourceProvider: string;
+    private _sdkVersion: string;
+    private _runtimeVersion: string;
+    private _os: string;
+    private _language: string;
+    private _ciIkey: string;
+    private _attach: string = Constants.StatsBeatAttach.sdk; // Default is SDK
+    private _features: number = Constants.StatsBeatFeature.NONE;
+    private _instrumentations: number = Constants.StatsBeatInstrumentation.NONE;
+
+    constructor(config?: Config) {
+        this._isInitialized = false;
+        this._config = config;
+        this._client = new TelemetryClient(STATS_BEAT_IKEY);
+    }
+
+    public enable(isEnabled: boolean) {
+        this._isEnabled = isEnabled;
+        if (this._isEnabled && !this._isInitialized) {
+            this._getCustomProperties();
+            this._isInitialized = true;
+        }
+        if (isEnabled) {
+            if (!this._handle) {
+                this._handle = setInterval(() => {
+                    this.trackStatsBeatMetrics()
+                }, this._collectionInterval);
+                this._handle.unref(); // Allow the app to terminate even while this loop is going on
+            }
+        } else {
+            if (this._handle) {
+                clearInterval(this._handle);
+                this._handle = null;
+            }
+        }
+    }
+
+    public isInitialized() {
+        return this._isInitialized;
+    }
+
+    public isEnabled() {
+        return this._isEnabled;
+    }
+
+    public setCodelessAttach() {
+        this._attach = Constants.StatsBeatAttach.codeless;
+    }
+
+    public addFeature(feature: Constants.StatsBeatFeature) {
+        this._features |= feature;
+    }
+
+    public removeFeature(feature: Constants.StatsBeatFeature) {
+        this._features &= ~feature;
+    }
+
+    public addInstrumentation(instrumentation: Constants.StatsBeatInstrumentation) {
+        this._instrumentations |= instrumentation;
```

Comment:
i would like to see some test for this to fully understand how you convert a list of strings to number (64-bit)?

Reply:
There is no list of strings, I use an enumerator describing the instrumentation and a number as value, number is always a power of 2.

Reply:
on the kusto side, you want to see the string right? number is not that intuitive. however, as long as you understand which number maps to which instrumentation, that's up to you. it's just not that clear to others.  Here is fine, but on the kusto end you might want to have another map to the string, and t would be nice if you can document it in the constants class to indicate which number maps to which instrumentation name.